src/ext.rs

```diff
@@ -0,0 +1,145 @@
+use core::{marker::PhantomData, task::Poll};
+
+use crate::{
+    api::{reply, request},
+    backend::{Backend, CoreOnly, Dispatch, NoId},
+    client::{ClientError, ClientImplementation, FutureResult, PollClient},
+    error::Error,
+    platform::{Platform, Syscall},
+    postcard_deserialize, postcard_serialize_bytes,
+    service::ServiceResources,
+    types::{Context, CoreContext},
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+
+pub trait Extension {
+    type Request: DeserializeOwned + Serialize;
+    type Reply: DeserializeOwned + Serialize;
+}
+
+/// Dispatches extension requests to custom backends.
+pub trait ExtensionDispatch<P: Platform>: Dispatch<P> {
+    /// The ID type for the extensions supported by this dispatch implementation.
+    type ExtensionId: TryFrom<u8, Error = Error>;
+
+    /// Executes an extension request using a backend or returns [`Error::RequestNotAvailable`][]
+    /// if it is not supported by the backend.
+    fn extension_request(
+        &mut self,
+        backend: &Self::BackendId,
+        extension: &Self::ExtensionId,
+        ctx: &mut Context<Self::Context>,
+        request: &request::Extension,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<reply::Extension, Error> {
+        let _ = (backend, extension, ctx, request, resources);
+        Err(Error::RequestNotAvailable)
+    }
+}
+
+impl<P: Platform> ExtensionDispatch<P> for CoreOnly {
+    type ExtensionId = NoId;
+}
+
+pub trait ExtensionImpl<E: Extension, P: Platform>: Backend<P> {
+    fn extension_request(
+        &mut self,
+        core_ctx: &mut CoreContext,
+        backend_ctx: &mut Self::Context,
+        request: &E::Request,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<E::Reply, Error>;
+
+    fn extension_request_serialized(
+        &mut self,
+        core_ctx: &mut CoreContext,
+        backend_ctx: &mut Self::Context,
+        request: &request::Extension,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<reply::Extension, Error> {
+        let request =
+            postcard_deserialize(&request.request).map_err(|_| Error::InvalidSerializedRequest)?;
+        let reply = self.extension_request(core_ctx, backend_ctx, &request, resources)?;
+        postcard_serialize_bytes(&reply)
+            .map(|reply| reply::Extension { reply })
+            .map_err(|_| Error::ReplySerializationFailure)
+    }
+}
+
+pub trait ExtensionId<E> {
+    type Id: Into<u8>;
+
+    const ID: Self::Id;
+}
+
+pub trait ExtensionClient<E: Extension>: PollClient {
+    fn id() -> u8;
+
+    fn extension<Rq, Rp>(&mut self, request: Rq) -> ExtensionResult<'_, E, Rp, Self>
```

Comment:
I think this could be done more elegantly, with less code, by using the `pub trait Extension` above (with a little polish like @sosthene-nitrokey's comments on Deserialize<'a>) and adding the method directly to `PollClient`, roughly as follows:

```rust
pub type ClientResult<'c, T, C> = core::result::Result<FutureResult<'c, T, C>, ClientError>;
pub type ClientExtensionResult<'c, T, C> = core::result::Result<FutureExtensionResult<'c, T, C>, ClientError>;

pub trait PollClient {
    fn request<T: From<crate::api::Reply>>(&mut self, req: impl Into<Request>) -> ClientResult<'_, T, Self>;
    fn poll(&mut self) -> core::task::Poll<core::result::Result<Reply, Error>>;
    fn syscall(&mut self);
    // new
    fn extension<E: Extension>(&mut self, req: E::Request) -> ClientExtensionResult<'_, E::Reply, Self>;
}
````
I know this is starting to contradict the ask to feature-flag extensions, but given we are including the "raw" Extension request in api.rs, if we can support handling it with minimal code we might as well keep it.
Then all the `PollClient::extension` has to do is serialize `E::Request` into the extension variant, and continue like `PollClient::poll`.

I even think we might be able to merge the two future result and client result types with suitable conversion traits. Or at the very least, we'd have not too much code to feature flag.